### PR TITLE
Move statistics page to Rails

### PR DESCRIPTION
**Pull request by timhabermaas**
## Todo
- [ ] Add view caches around each table and around the entire page.
- [x] Get rid of "extended" names (`Kanneti Sae Han (คันธ์เนตี แซ่ห่าน)`) in tables (see also #186).
- [x] Fix headlines / helper texts, move them to the appropriate statistic model.
- [x] Move the rest of the statistics over to Rails
- [ ] Add unit tests for the statistic models + one feature smoke test for the `/statistics` page
- [ ] Replace `TimeTd` with `SolveTd`

### CODE REVIEW

**Comment by timhabermaas** [comment]
For the record: I didn't copy the query for the `top 100 results` statistic from the PHP code, but rewrote it using pure SQL (`UNION` and subqueries). It seems to be correct, but if it isn't, I messed up.

File: WcaOnRails/app/views/statistics/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+              <% elsif statistic.subtitle %>
+                <span style="color:#999">(<%= statistic.subtitle %>)</span>
+              <% end %>
+            </td>
+          </tr>
+
+          <!-- shouldn't this be in a thead? -->
+          <tr>
+            <% statistic.headers.each do |header| %>
+              <% if header.respond_to?(:render) %>
```
Comment (timhabermaas) [inline WcaOnRails/app/views/statistics/index.html.erb]: Note to self: This shouldn't be necessary.

**Comment by timhabermaas** [comment]
Once #381 is merged I should probably fix the competition link in `CompetitionTd`.

**Comment by timhabermaas** [comment]
I'm finished with porting all statistics to the Rails page. The performance is unacceptable, though. The entire page takes about a minute to render and individual statistics up to 10s. Therefore view caching doesn't help much. I'll experiment with keeping the tables in memory and refreshing them in the background (using threads) if the they aren't up-to-date anymore.

**Comment by jfly** [comment]
Wow, that's awesome, Tim! It might be time to build some infrastructure for background jobs. I'll send out an email with my thoughts soon.

File: WcaOnRails/app/assets/stylesheets/links.scss [hunk 2]
```diff
@@ -14,5 +14,6 @@ form[target=_blank] button[type=submit] {
 }
 
 :target {
-  background-color: yellow;
+  // TODO: why?
```
Comment (jfly) [inline WcaOnRails/app/assets/stylesheets/links.scss]: So you can tell what I'm linking you to here: https://www.worldcubeassociation.org/regulations/#Z3

File: WcaOnRails/app/helpers/path_helper.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+module PathHelper
```
Comment (jfly) [inline WcaOnRails/app/helpers/path_helper.rb]: Awesome! I apologize for not doing this forever ago.

File: WcaOnRails/app/models/statistics.rb [hunk 4]
```diff
@@ -0,0 +1,167 @@
+module Statistics
+  TextTd = Struct.new(:text) do
+    def render
+      "<td>#{text}</td>".html_safe
```
Comment (larspetrus) [inline WcaOnRails/app/models/statistics.rb]: So why manually generate HTML in a model?

Shouldn't this be view helper code?

File: WcaOnRails/app/views/statistics/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          <h3>
+            <%= statistic.name %>
+            <% if statistic.subtitle %>
+              <small><%= statistic.subtitle %></small>
+            <% end %>
+          </h3>
+        </div>
+        <% tables = statistic.tables %>
+        <% tables.each do |t| %>
+          <div class="col-md-<%= 12 / tables.size %>">
```
Comment (jfly) [inline WcaOnRails/app/views/statistics/index.html.erb]: Nitpick, but wow, this took me embarassingly long to figure out... `.size`, `.count`, and `.length` all do the same thing in ruby? I think we've been using `.length` elsewhere, how do you feel about being consistent about using length everywhere?

**Comment by jfly** [comment]
This looks pretty good to me, I like the way you split up all the statistics. I didn't actually read through any of the SQL, I'm assuming it came from the old code and doesn't need a careful eye.

File: WcaOnRails/app/views/layouts/php_land.html.erb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    <link rel="shortcut icon" href="/results/images/wca.ico">
+
+    <script async="" src="//www.google-analytics.com/analytics.js"></script><script src="//ajax.googleapis.com/ajax/libs/jquery/2.1.0/jquery.min.js?v=201409231900"></script><script src="/results/js/ga.js?v=201409231900"></script><script src="/results/js/bootstrap.min.js?v=201409231900"></script><script src="/results/js/bootstrap-hover-dropdown.min.js?v=201409231900"></script><link rel="stylesheet" href="/results/style/bootstrap.min.css?v=201409231900" type="text/css"><link rel="stylesheet" href="/results/style/navbar-static-top.css?v=201409231900" type="text/css"><link rel="stylesheet" href="/results/style/general.css?v=201409231900" type="text/css"><link rel="stylesheet" href="/results/style/pageMenu.css?v=201409231900" type="text/css"><link rel="stylesheet" href="/results/style/tables.css?v=201409231900" type="text/css"><link rel="stylesheet" href="/results/style/links.css?v=201409231900" type="text/css">
+
+  </head>
+  <body data-rails-controller-name="<%= controller_name %>" data-rails-controller-action-name="<%= action_name %>">
+    <!-- Static navbar -->
+    <%= render 'layouts/navigation' %>
```
Comment (jfly) [inline WcaOnRails/app/views/layouts/php_land.html.erb]: Hmm... does our new navbar work without our special css? Is it a lot of work to get this working with our shared layout?

**Comment by jfly** [comment]
Closing due to inactivity. See our new policy: [PRs that have gone 3 months without activity are subject to being closed.](https://github.com/thewca/worldcubeassociation.org/wiki/Merging-and-deploying)

**Pull request closed without merging**